trieste/acquisition/function/function.py

```diff
@@ -1034,6 +1035,595 @@ def __call__(self, x: TensorType) -> TensorType:
+        )
+
+        # Unpack and mean shape
+        B, Q = mean.shape
+
+        # Compute b and m tensors
+        b, m = self.compute_bm(
+            mean=mean,
+            threshold=threshold,
+        )  # (B, Q, Q), (B, Q, Q)
+
+        # Compute Sigma
+        Sigma = self.compute_Sigma(covariance=covariance)  # (B, Q, Q, Q)
+
+        # Reshape all tensors, for batching
+        b_reshaped = tf.reshape(b, (B * Q, Q))
+        m_reshaped = tf.reshape(m, (B * Q, Q))
+        Sigma_reshaped = tf.reshape(Sigma, (B * Q, Q, Q))
+
+        # Compute p tensor
+        p = self.compute_p(
+            m_reshaped=m_reshaped,
+            b_reshaped=b_reshaped,
+            Sigma_reshaped=Sigma_reshaped,
+            mvn_cdf=mvn_cdf,
+        )
+
+        # Compute c
+        c = self.compute_c(
+            m_reshaped=m_reshaped,
+            b_reshaped=b_reshaped,
+            Sigma_reshaped=Sigma_reshaped,
+        )  # (B*Q, Q, Q-1)
+
+        # Compute Sigma_i
+        R = self.compute_R(
+            Sigma_reshaped=Sigma_reshaped,
+        )  # (B*Q, Q, Q-1, Q-1)
+
+        # Compute Q-1 multivariate CDFs
+        Phi_mvn_cdfs = self.compute_Phi(
+            c=c,
+            R=R,
+            mvn_cdf=mvn_cdf,
+        )
+
+        # Compute univariate pdfs
+        S_diag = tf.linalg.diag_part(Sigma)
+        normal = tfp.distributions.Normal(loc=m, scale=S_diag ** 0.5)
+        uvn_pdfs = tf.math.exp(normal.log_prob(b))  # (B, Q, Q)
+
+        Sigma_diag = tf.linalg.diag_part(tf.transpose(Sigma, perm=[0, 2, 1, 3]))
+        Sigma_diag = tf.transpose(Sigma_diag, perm=[0, 2, 1])
+
+        T = tf.tile(threshold[:, None], (1, Q))
+
+        mean_T_term = (mean - T) * p
+
+        # Compute inner sum
+        sum_term = tf.reduce_sum(
+            Sigma_diag * uvn_pdfs * Phi_mvn_cdfs,
+            axis=2,
+        )
+
+        # Compute outer sum
+        expected_improvement = tf.reduce_sum(mean_T_term + sum_term, axis=1)
+
+        return expected_improvement
+
+    @tf.function
+    def __call__(self, x: TensorType) -> TensorType:
+        """Computes the accurate approximation of the multi-point expected
+        improvement.
+
+        :param x: Tensor of shape ***.
+        :returns ei: Tensor of shape (B,), expected improvement.
+        """
+
+        if not hasattr(self, "_mvn_cdf"):
+            self._mvn_cdf = make_mvn_cdf(samples=self._samples)
+
+        mean, covariance = self._model.predict_joint(x)  # type: ignore
+
+        mean = mean[:, :, 0]
+        covariance = covariance[:, 0, :, :]
+        covariance = covariance + 1e-6 * tf.eye(
+            covariance.shape[-1],
+            dtype=covariance.dtype,
+        )[None, :, :]
+
+        threshold = tf.tile(self._eta, (mean.shape[0],))
+
+        ei = self.compute_batch_expected_improvement(
+            mean=-mean,
+            covariance=covariance,
+            threshold=-threshold,
+            mvn_cdf=self._mvn_cdf,
+        )[:, None]
+
+        return ei
+
+
```

Comment:
@uri-granta  are you sure this shoudnt be another file?

Reply:
I originally had this on a separate utils file and @uri-granta suggested I should put all functions which are specific to the batch expected improvement on this file, since it is not (and will not) be used by anything else. By contrast, the multivariate normal CDF may be used in future parts of the codebase so it makes more sense for it to be in a separate utils file.